src/EntityId/PrefixingEntityIdParser.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+namespace Wikibase\DataModel\Services\EntityId;
+
+use Wikibase\DataModel\Entity\EntityId;
+use Wikibase\DataModel\Entity\EntityIdParser;
+
+/**
+ * EntityIdParser that adds a fixed prefix and parses resulting string as an EntityId.
+ * This can be used to prefix IDs of entities coming from a foreign repository
+ * with the repository name to avoid clashes with IDs of local entities.
+ *
+ * @since 3.7
+ *
+ * @license GPL-2.0+
+ */
+class PrefixingEntityIdParser implements EntityIdParser {
+
+	/**
+	 * @var string
+	 */
+	private $prefix;
+
+	/**
+	 * @var EntityIdParser
+	 */
+	private $idParser;
+
+	/**
+	 * @param string $prefix Prefix to be added
+	 * @param EntityIdParser $idParser
+	 */
+	public function __construct( $prefix, EntityIdParser $idParser ) {
+		$this->prefix = $prefix;
+		$this->idParser = $idParser;
+	}
+
+	/**
+	 * Adds the prefix to the given string and parses the resulting string as an EntityId.
+	 * If the string is already prefixed, the prefix is not duplicated (this check is case sensitive).
+	 *
+	 * @param string $idSerialization
+	 * @return EntityId
+	 */
+	public function parse($idSerialization) {
+		// Do not add prefix if $idSerialization already starts with it
+		if ( strncmp( $idSerialization, $this->prefix, strlen( $this->prefix ) ) !== 0 ) {
+			$idSerialization = $this->prefix . $idSerialization;
```

Comment:
Indeed. I thought of this parser as a kind of generic parser that adds some prefix and not something foreign-id-specific. If we follow this path the code using the parser must ensure that the colon is present in a prefix passed to the parser.
This might be a more generic approach but on the other hand I believe the only use of such parser we now can think of would be adding a foreign repo prefix, wouldn't it? So it indeed seems reasonable to simply have `ForeignEntitiyIdPrefixingParser` (better name needed).
In the latter case I would also prefer not to hard code the colon in this parser. That said I agree with @JeroenDeDauw and thinking of having a static call just to avoid simple a string concat makes me pull out my hair.

>  In the PR to EntityId I can't find how these things are called. "Prefixes" is an accurate technical term, though perhaps not the best choice of naming for something with more specific meaning in our domain.

At some point we have been leaning towards calling these "foreign repository names", so something along this lines could probably be part of the name of the parser instead of "prefix".